src/ft60x.rs

```diff
@@ -97,37 +105,131 @@ impl FT60x {
+        let (full_buffer_tx, full_buffer_rx) = sync_channel(10);
+
+        let mut thread_fn = move || {
+            self.set_streaming_mode().unwrap();
+
+            let blocksize: usize = 32 * 1024; // 32 Kb seems to be the sweet spot for the ft601
+
+            let mut async_group_buffer: Vec<(Vec<u8>, AsyncGroup)> = Vec::new();
+
+            let mut outstanding = 0;
+            for mut current_buffer in empty_buffer_rx.iter() {
+                let mut current_async_group = AsyncGroup::new(&self.context);
+                for chunk in unsafe {
+                    std::mem::transmute::<&mut _, &'static mut Vec<u8>>(&mut current_buffer)
+                }
+                .chunks_mut(blocksize)
+                {
+                    // The FT60x doesn't seem to like too many outstanding requests
+                    if outstanding > 500 {
+                        let mut to_ship = None;
+                        for (i, async_group) in async_group_buffer
+                            .iter_mut()
+                            .map(|v| &mut v.1)
+                            .chain(once(&mut current_async_group))
+                            .enumerate()
+                        {
+                            match async_group.wait_any() {
+                                Ok(mut transfer) => {
+                                    ensure!(
+                                        transfer.buffer().len() == transfer.actual().len(),
+                                        "FT60x did not return enough data. requested {} got {}",
+                                        transfer.buffer().len(),
+                                        transfer.actual().len()
+                                    );
+                                    outstanding -= 1;
+                                    break;
+                                }
+                                Err(rusb::Error::NotFound) => {
+                                    assert_eq!(to_ship, None);
+                                    assert_eq!(i, 0);
+                                    to_ship = Some(i);
+                                }
+                                Err(_) => panic!(),
+                            }
+                        }
+
+                        if let Some(i) = to_ship {
+                            if i < async_group_buffer.len() {
+                                full_buffer_tx.send(async_group_buffer.remove(i).0).unwrap();
+                            }
+                        }
+                    }
+
+                    current_async_group.submit(Transfer::bulk(
+                        &self.device,
+                        0x82,
+                        chunk,
+                        Duration::new(1, 0),
+                    ));
+                    outstanding += 1;
+                }
+                async_group_buffer.push((current_buffer, current_async_group));
+            }
```

Comment:
This needs to wait until all (potential) transactions for the buffers / `AsyncGroup`s still in `async_group_buffer` are done.